### PR TITLE
chore(nexus): reorganize filestream into a package

### PR DESCRIPTION
**Pull request by raubitsj**
Description
-----------
Objective of this PR: Reorganize and rename methods but do not impact functionality in this PR

Follow on PRs:
- more testing
- better handling of updates with multiple data types
- more cleanup of naming of internal structures

### <samp>🤖 Generated by Copilot at de5d195</samp>

This pull request introduces a new `filestream` package that handles streaming data to the W&B backend filestream service. It refactors the existing `server` package to use the `filestream` package for sending data, and updates the tests and imports accordingly. It also adds functional options to the `filestream` package to customize its behavior.

Testing
-------
golang tests

### CODE REVIEW

File: nexus/pkg/server/sender.go [hunk 1]
```diff
@@ -362,11 +364,11 @@ func (s *Sender) checkAndUpdateResumeState(run *service.RunRecord) error {
 	}
 
 	if s.resumeState.FileStreamOffset == nil {
-		s.resumeState.FileStreamOffset = make(map[chunkFile]int)
+		s.resumeState.FileStreamOffset = make(map[fs.ChunkFile]int)
```
Comment (kptkin) [inline nexus/pkg/server/sender.go:367]: nit: but you already have a type for this map... should we just use it?

File: nexus/pkg/server/sender.go [hunk 2]
```diff
@@ -175,12 +176,13 @@ func (s *Sender) sendRunStart(_ *service.RunStartRequest) {
...
-				s.fileStream.SetOffset(k, v)
-			}
-		}
+		s.fileStream = fs.NewFileStream(
+			fs.WithPath(fsPath),
+			fs.WithSettings(s.settings),
+			fs.WithLogger(s.logger),
+			fs.WithHttpClient(NewRetryClient(s.settings.GetApiKey().GetValue(), s.logger)),
+			fs.WithOffsets(s.resumeState.FileStreamOffset),
+		)
```
Comment (kptkin) [inline nexus/pkg/server/sender.go:185]: big improvement!

File: nexus/pkg/server/sender.go [hunk 3]
```diff
@@ -26,7 +27,7 @@ const (
 )
 
 type ResumeState struct {
-	FileStreamOffset map[chunkFile]int
+	FileStreamOffset fs.FileStreamOffsetMap
 	Error            service.ErrorInfo
```
Comment (kptkin) [inline nexus/pkg/server/sender.go:31]: i though i got rid of the Error field in the resumeState (i don't think it is needed anymore)

File: nexus/pkg/server/clients.go [hunk 4]
```diff
@@ -28,8 +28,8 @@ func (t *authedTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 	return t.wrapped.RoundTrip(req)
```
Comment (kptkin) [inline nexus/pkg/server/clients.go:28]: eventually we probably want to organize this differently, but i think it's ok for now

File: nexus/pkg/filestream/loop_transmit.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	fileName string
+	fileData *chunkLine
+	Exitcode *int32
+	Complete *bool
+}
+
+type chunkLine struct {
+	chunkType ChunkFile
+	line      string
+}
```
Comment (kptkin) [inline nexus/pkg/filestream/loop_transmit.go:25]: maybe re-think these types, it was quite awkward in the send function when we combining the chunks

File: nexus/pkg/filestream/loop_transmit.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		if fs.stageExitChunk != nil {
+			fs.sendChunkList([]chunkData{*fs.stageExitChunk})
+		}
+	}
+}
+
+func (fs *FileStream) sendChunkList(chunks []chunkData) {
+	var lines []string
+	var complete *bool
+	var exitcode *int32
```
Comment (kptkin) [inline nexus/pkg/filestream/loop_transmit.go:101]: can we use a new chunk instead it already has all these fields.

File: nexus/pkg/filestream/filestream.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+func WithOffsets(offsetMap FileStreamOffsetMap) FileStreamOption {
+	return func(fs *FileStream) {
+		for k, v := range offsetMap {
+			fs.offsetMap[k] = v
+		}
+	}
+}
+
+func NewFileStream(opts ...FileStreamOption) *FileStream {
```
Comment (kptkin) [inline nexus/pkg/filestream/filestream.go:126]: nice improvement with the options!

**Review by kptkin — APPROVED** [review]

File: nexus/pkg/server/sender.go (hunk 1 above)
Comment (raubitsj) [reply nexus/pkg/server/sender.go:367]: yes. when i did that search and replace i was planning on coming back to that.

File: nexus/pkg/server/sender.go (hunk 3 above)
Comment (raubitsj) [reply nexus/pkg/server/sender.go:31]: i was leaving it for now, but i think we can do another pass for some error stuff.

File: nexus/pkg/server/clients.go (hunk 4 above)
Comment (raubitsj) [reply nexus/pkg/server/clients.go:28]: yes.  client changes are needed for sure.

File: nexus/pkg/filestream/loop_transmit.go (hunk 5 above)
Comment (raubitsj) [reply nexus/pkg/filestream/loop_transmit.go:25]: yes, at least they are more private now and tied to the "transmit". but this was getting more attention in the PR which was going to make sending different types of data better.

File: nexus/pkg/filestream/loop_transmit.go (hunk 6 above)
Comment (raubitsj) [reply nexus/pkg/filestream/loop_transmit.go:101]: yes, this was a reorg, but not changing the code too much, the next change is redoing sendChunkList

File: nexus/pkg/server/sender.go (hunk 1 above)
Comment (raubitsj) [reply nexus/pkg/server/sender.go]: fixed!

**Pull request merged**